Sheshuk/flavor transformations with flavor matrices

Sheshuk opened the pull request:
To be merged after #350 

In this PR I do the following:

1. Take the FlavorTransformation classes, implemented by @jpkneller in #308
2. Switch to using the `FlavorMatrix` class everywhere in `get_probabilities`
3. Move the `Three/FourFlavorTransformation.Pmf_HighDensityLimit` methods to `Three/FourFlavorMixingParameters.Pmf_HighDensityLimit` - I think they belong next there, next to `VacuumMixingMatrix`
4. ...refactor the class hierarchy, to make a more clear user interface...

sybenzvi commented:
Comment from @Sheshuk (2024-08-13): some classes are implementing all flavor transformations. Others are handling transformations only in certain environments, e.g., in Earth or in vacuum. As a result, the matrix outputs are not consistent or equivalent across all classes (e.g., flavor-to-flavor, flavor-to-mass, mass-to-mass, etc.).

This needs to be checked carefully in the code to ensure that mixing operations between the classes is not allowed if physically the transformations should not be mixed. All classes inherit from `FlavorTransformation` and the class hierarchy needs to be refactored to prevent unphysical combinations.

sybenzvi commented:
A dumb observation @Sheshuk: while approving PR #350, I noticed the `snewpy.utils` submodule has just one function. While it is called in `flux.py` and `flavor.py`, but really it's quite specific to flavor transformations.

While I could live with it, I don't love having a generic "utils" module that is destined to be a dumping ground for random calculations. Can we move the function into `expand_dimensions_to` into `flavor.py` and delete `utils.py`?

sybenzvi commented:
@Sheshuk, I added support for `BEMEWS` to `snewpy.flavor_transformation`. It does work but I noticed a few other problems with the classes in this submodule:
- Most of the classes assumed the `MixingParameters` objects had a property called `mix_params` but it's apparently `mix_pars` now.
- Most of the `SNTransformation` and `EarthTransformation` children expect a private `mix_pars` member. This was not being defined so I added `mix_params` to the parent class constructor.
- Nearly all of the unit tests are broken because of attempted conversions of `ThreeFlavor_MassBasis` to `TwoFlavor` which are invalid. Presumably this is a holdover from an older version of the code that did not handle the mass and flavor bases separately.

Sheshuk commented:
Finally I think this is ready for review. 
Here is a summary so we can discuss it.

## What I did
1. Separated "concrete" flavor transformation classes into a tree, so that SN, Vacuum and Earth transformations are different classes, and distinct from the "full" transformations (children of `FlavorTransformation` class)
![classes](https://github.com/user-attachments/assets/82443b1e-f57f-4b4f-a9b4-c98ba7c2ac7e)
2. Made `flavor_transformation` a nested package instead of single file (so `in_sn`,`in_vacuum` and `in_earth` contain corresponding transforms)
3. Made each transformation produce a `FlavorMatrix` object, which is very convenient for accessing, multiplication etc. instead of a plain `np.array`
4. Updated the tests:  convert the transformation matrix `TwoFlavor>>P_ff>>TwoFlavor` and use the same 2Flavor formulas in the tests as before (they still need update in the future)
5. Added new tests: using the FlavorMatrix we can more easily check the matrix form against the equations in snewpy v2 draft, for example: https://github.com/SNEWS2/snewpy/blob/d8664a617d41dc8944601603bc3718334997634b/python/snewpy/test/test_04_xforms.py#L312-L319
![image](https://github.com/user-attachments/assets/c43acc0b-91f4-4669-97c4-c180137ea631)
6. User interface: since now we have all the concrete classes like `AdiabaticMSW` as building blocks of the `TransformationChain` the user would have to build the desired chain from scratch... So for backward compatibility I declared the pre-defined TransformationChains with the same names, as their main components:
https://github.com/SNEWS2/snewpy/blob/d8664a617d41dc8944601603bc3718334997634b/python/snewpy/flavor_transformation/__init__.py#L44-L51
I'm not sure this is the best way to go. We will probably need to revisit this.
7. Added some basic documentation, we'll need to revisit it as well, once the interface is finalized.

## Next steps (to be done in separate PRs):
- [ ] Update tests with 3flavor formulas
- [ ] Finalize the user interface
- [ ] Update the documentation
- [ ] SNOSHEWS support needs to be tested separately
- [ ] Implement 3flavor rate calculation

Sheshuk commented:
One more thing: I had to swap the formulas for NMO and IMO for NeutrinoDecay in the tests: 09735b53cbfe115b2859ebb0584d425d812236d7 I'm not sure this is correct. we need a cross-check

JostMigenda reviewed (commented):
Individual comments are at the appropriate line in the code.

Regarding the code arrangement in `flavor_transformation/`:
In addition to the `in_*` submodules (which makes sense), there are three fairly generic submodules (`__init__.py`, `base.py` and `transforms.py`) and it’s not obvious to me what code goes where. My intuition would probably be to drop `transforms.py` by putting `NoTransformation`/`CompleteExchange`/`ThreeFlavorDecoherence` into `__init__.py` (alongside the backwards-compatible transformation chains) and moving `FlavorTransformation` and `TransformationChain` into `base.py`?

Finally, I added a commit cleaning up unused imports.

Sheshuk commented:
@JostMigenda thanks a lot for your review!

> * I can see `in_earth.NoEarthMatter` and `in_vacuum.NoVacuumTransformation`, but no equivalent `in_sn.NoSNTransformation`? And that’s presumably why `in_sn` is a required argument to a `TransformationChain`, while `in_vacuum` and `in_earth` are optional? 

Yes, it is correct. I also don't like it being asymmetric, so if you have any idea of making it more clear for the user, please tell.
Another misleading thing is that we can't reproduce `NoTransformation` in `TransformationChain` - even if we completely ignore the SN matter and other effects, the flavor states are transformed to mass ones $P_{\alpha\to i}$ and then back $P_{i\to\beta}$, but because of the decoherence of mass states, these transformation matrices don't produce a unit matrix $I_{\alpha\beta}$.

> 
> * I’d like to rename `MSWEffect`, since that name sounds so generic that novice users would likely expect that to be the basic MSW case. Something like `CustomMSW` or `AdvancedMSW` would be clearer (and more consistent with `AdiabaticMSW`, `NonAdiabaticMSW` and their four-flavour equivalents, which all drop the “Effect” from the class name).

I agree, `CustomMSW` would be better here.
I kept all the class names initially defined by Jim in his branch, and didn't change anything because I'm not sure I have enough understanding of underlying physics 😃 
But it will be good if we can come with a more clear naming convention in a discussion.

But probably we should do the renaming in a separate PR, it will be rather straightforward.

>  The TransformationChain instances are currently saving the transforms in two different places; can we simplify that?

Sure. I can make `in_sn, in_vacuum, in_earth` properties, accessing the elements of `self.transforms`. Or vice versa. Would that be better?

> In addition to the `in_*` submodules (which makes sense), there are three fairly generic submodules (`__init__.py`, `base.py` and `transforms.py`) and it’s not obvious to me what code goes where. My intuition would probably be to drop `transforms.py` by putting `NoTransformation`/`CompleteExchange`/`ThreeFlavorDecoherence` into `__init__.py` (alongside the backwards-compatible transformation chains) and moving `FlavorTransformation` and `TransformationChain` into `base.py`?

That makes sense, thank you, I'll do that!

The only classes standing out here are `Three/FourFlavorTransformation` in `base.py`, which are not base classes per se, but a small functionality, providing a way to say "this class should have Three(Four)FlavorMixingParameters,no matter what. If you try to change the parameters, just change the values and don't change their type". 
I'm not sure this is the most elegant way to implement it. If you have any ideas, please share

Sheshuk commented:
> In addition to the `in_*` submodules (which makes sense), there are three fairly generic submodules (`__init__.py`, `base.py` and `transforms.py`) and it’s not obvious to me what code goes where. My intuition would probably be to drop `transforms.py` by putting `NoTransformation`/`CompleteExchange`/`ThreeFlavorDecoherence` into `__init__.py` (alongside the backwards-compatible transformation chains) and moving `FlavorTransformation` and `TransformationChain` into `base.py`?

On second thought, this makes a circular import:
`TransformationChain` needs `NoVacuumTransformation`,`NoEarthMatter` for the default values (and `SNTransformation`,`VacuumTransformation` and `EarthTransformation` for type hints), which depend on `FlavorTransformation` base class. So we can't put `FlavorTransformation` in `base.py`.
We can put it together with the rest transorms in `__init__.py`, but that will bascially mean that we migrated our `transforms.py` into `__init__.py`. I personally don't like when there is a lot of code in `__init__.py` except for some small definitions of the objects to be imported etc.

JostMigenda commented:
> > * I can see `in_earth.NoEarthMatter` and `in_vacuum.NoVacuumTransformation`, but no equivalent `in_sn.NoSNTransformation`? And that’s presumably why `in_sn` is a required argument to a `TransformationChain`, while `in_vacuum` and `in_earth` are optional?
> 
> Yes, it is correct. I also don't like it being asymmetric, so if you have any idea of making it more clear for the user, please tell. Another misleading thing is that we can't reproduce `NoTransformation` in `TransformationChain` - even if we completely ignore the SN matter and other effects, the flavor states are transformed to mass ones P α → i and then back P i → β , but because of the decoherence of mass states, these transformation matrices don't produce a unit matrix I α β .

Hm … good point; I can see it being very confusing if `TransformationChain(NoSNTransformation, NoVacuumTransformation, NoEarthMatter)` does not match `NoTransformation`. @jpkneller, do you have any idea how to make it clearer? Or should we just live with that asymmetry and document it well?

> > * I’d like to rename `MSWEffect`, since that name sounds so generic that novice users would likely expect that to be the basic MSW case. Something like `CustomMSW` or `AdvancedMSW` would be clearer (and more consistent with `AdiabaticMSW`, `NonAdiabaticMSW` and their four-flavour equivalents, which all drop the “Effect” from the class name).
> 
> I agree, `CustomMSW` would be better here. I kept all the class names initially defined by Jim in his branch, and didn't change anything because I'm not sure I have enough understanding of underlying physics 😃 But it will be good if we can come with a more clear naming convention in a discussion.
> 
> But probably we should do the renaming in a separate PR, it will be rather straightforward.

Yep, let’s do that! I’ve created #366 to keep track of it.

> > The TransformationChain instances are currently saving the transforms in two different places; can we simplify that?
> 
> Sure. I can make `in_sn, in_vacuum, in_earth` properties, accessing the elements of `self.transforms`. Or vice versa. Would that be better?

It would certainly eliminate any risk of bugs where we change e.g. `self.in_sn` but forget to update `self.transforms`, yes.
After thinking about it a bit longer yesterday, however, I think making `self.transforms` a `namedtuple` would be even better?

```Python
>>> from collections import namedtuple
>>> Transformations = namedtuple('Transformations', 'in_sn, in_vacuum, in_earth')
>>> transforms = Transformations('AdiabaticMSW', 'NoVacuumTransformation', 'NoEarthMatter')
>>> for t in transforms:
...     print(t)
... 
AdiabaticMSW
NoVacuumTransformation
NoEarthMatter
>>> transforms.in_vacuum
'NoVacuumTransformation'
```
In particular, the current design doesn’t make it clear that iterating over `self.transforms` and accessing `self.in_*` is equivalent; whereas the namedtuple makes that obvious.
(Personally, I also think that `self.transforms.in_vacuum`, while a bit more verbose, is more readable than `self.in_vacuum`; but that’s more subjective.)

> The only classes standing out here are `Three/FourFlavorTransformation` in `base.py`, which are not base classes per se, but a small functionality, providing a way to say "this class should have Three(Four)FlavorMixingParameters,no matter what. If you try to change the parameters, just change the values and don't change their type". I'm not sure this is the most elegant way to implement it. If you have any ideas, please share

I guess they are more of a [mix-in](https://en.wikipedia.org/wiki/Mixin) than a base class; but I’m fine with that.


> > In addition to the `in_*` submodules (which makes sense), there are three fairly generic submodules (`__init__.py`, `base.py` and `transforms.py`) and it’s not obvious to me what code goes where. My intuition would probably be to drop `transforms.py` by putting `NoTransformation`/`CompleteExchange`/`ThreeFlavorDecoherence` into `__init__.py` (alongside the backwards-compatible transformation chains) and moving `FlavorTransformation` and `TransformationChain` into `base.py`?
> 
> On second thought, this makes a circular import: 

Oh, good point!

After thinking about it for a while, I think the main issue is the `TransformationChain`—it’s a high-level class which depends on the `SN/Vacuum/EarthTransformation` classes, so it shouldn’t be lumped in with the base classes. With that in mind, I would suggest restructuring as follows:
* One submodule for all concrete transformation presets (both phenomenological ones (`NoTransformation`/`CompleteExchange`/`ThreeFlavorDecoherence`) and the backwards-compatible `TransformationChain`s)
* One submodule for base classes and mix-ins (i.e. `FlavorTransformation` and `Three/FourFlavorTransformation`)
* A separate submodule for the `TransformationChain` class

The first of these should probably be `__init__.py` (where the backwards-compatible `TransformationChain`s already live); the second should be `base.py`; and I’d probably rename the last one from `transforms.py` to `TransformationChain.py` for clarity.
Since I already went through the whole process, I’ll post a separate PR in a minute with these changes for you to look at (and merge into this PR branch, if you’re happy with it).

Sheshuk commented:
> After thinking about it a bit longer yesterday, however, I think making `self.transforms` a `namedtuple` would be even better?

Good idea, that's perfect use case for that. I also like `self.transforms.in_vacuum` - it's always nice when the code reads as a natural language.

I'll do it now.

Sheshuk commented on python/snewpy/flavor_transformation/base.py:
```diff
@@ -0,0 +1,41 @@
+        """
+        pass
+
+    def apply(self, flux: Container) -> Container:
+        r"""Apply this transformation to the given flux, return transformaed flux"""
```
You're right, thats counterintuitive. Of course what I mean was `apply_to`, I will change the name. Should I do it here, or in #363?

JostMigenda replied:
I think logically it belongs in this PR; so I’d do it here unless the merge/rebase of #363 becomes too painful.

JostMigenda commented:
Force-pushed an updated version of the branch, which is now rebased on top of the merged `main`+`release_v2.0` branch; as discussed with @Sheshuk on Slack.

Apologies for any complications caused by this.

jpkneller commented:
> > * I can see `in_earth.NoEarthMatter` and `in_vacuum.NoVacuumTransformation`, but no equivalent `in_sn.NoSNTransformation`? And that’s presumably why `in_sn` is a required argument to a `TransformationChain`, while `in_vacuum` and `in_earth` are optional?
> 
> Yes, it is correct. I also don't like it being asymmetric, so if you have any idea of making it more clear for the user, please tell. Another misleading thing is that we can't reproduce `NoTransformation` in `TransformationChain` - even if we completely ignore the SN matter and other effects, the flavor states are transformed to mass ones P α → i and then back P i → β , but because of the decoherence of mass states, these transformation matrices don't produce a unit matrix I α β .

From a physics perspective there is no symmetry between the three parts of a transformation chain: the neutrinos have to pass through the supernova but they don't have to do anything in the vacuum (except decohere), or pass through the Earth. I would prefer that we not provide a in_sn.NoSNTransformation prescription so that the user cannot create an empty TransformationChain object that they might think would be equivalent to NoTransformation. But if you insist, the only way I can see how to do it would be to check that all three parts of the chain are None and if so, use NoTransformation. 
 

> > * I’d like to rename `MSWEffect`, since that name sounds so generic that novice users would likely expect that to be the basic MSW case. Something like `CustomMSW` or `AdvancedMSW` would be clearer (and more consistent with `AdiabaticMSW`, `NonAdiabaticMSW` and their four-flavour equivalents, which all drop the “Effect” from the class name).
> 
> I agree, `CustomMSW` would be better here. I kept all the class names initially defined by Jim in his branch, and didn't change anything because I'm not sure I have enough understanding of underlying physics 😃 But it will be good if we can come with a more clear naming convention in a discussion.

AdvancedMSW is fine with me, or we could adopt NumericalMSW

jpkneller commented:
I noticed that ThreeFlavorDecoherence is described as Phenomenological transformation and placed in __init__.py. It really should be a SN transformation. It's true that Earth matter can't change the decoherence but vacuum transformations could e.g. one of the mass states decays.

JostMigenda commented:
> From a physics perspective there is no symmetry between the three parts of a transformation chain: the neutrinos have to pass through the supernova but they don't have to do anything in the vacuum (except decohere), or pass through the Earth. I would prefer that we not provide a in_sn.NoSNTransformation prescription so that the user cannot create an empty TransformationChain object that they might think would be equivalent to NoTransformation. But if you insist, the only way I can see how to do it would be to check that all three parts of the chain are None and if so, use NoTransformation.

I just wanted to make sure that this asymmetry is deliberate and not an oversight.
It clearly is and we have a good explanation for it, so I’m happy to leave it as is. (Though this might be a good point to highlight in the docs, in case any users have a similar question.)

> I noticed that ThreeFlavorDecoherence is described as Phenomenological transformation and placed in `__init__.py`. It really should be a SN transformation. It's true that Earth matter can't change the decoherence but vacuum transformations could e.g. one of the mass states decays.

My impression was that `NoTransformation`, `CompleteExchange` and `ThreeFlavorDecoherence` are typically used as “black box transformations”—where researchers are interested in some other aspect and use these extreme scenarios to get a rough idea of how flavor transformations could affect studies of that other aspect. So it makes sense to me to keep these three separate?

However, if you think that any users might want to build a `TransformationChain` out of `ThreeFlavorDecoherence` in the SN plus additional transformations in vacuum/earth, then we could turn TFD into a SNTransformation.
In that case, would `TransformationChain(in_sn=ThreeFlavorDecoherence())` automatically result in the same flavour equilibration that we get from the current implementation of TFD? If not, that might be very confusing.

jpkneller commented:
> > From a physics perspective there is no symmetry between the three parts of a transformation chain: the neutrinos have to pass through the supernova but they don't have to do anything in the vacuum (except decohere), or pass through the Earth. I would prefer that we not provide a in_sn.NoSNTransformation prescription so that the user cannot create an empty TransformationChain object that they might think would be equivalent to NoTransformation. But if you insist, the only way I can see how to do it would be to check that all three parts of the chain are None and if so, use NoTransformation.
> 
> I just wanted to make sure that this asymmetry is deliberate and not an oversight. It clearly is and we have a good explanation for it, so I’m happy to leave it as is. (Though this might be a good point to highlight in the docs, in case any users have a similar question.)

Will do. 


> > I noticed that ThreeFlavorDecoherence is described as Phenomenological transformation and placed in `__init__.py`. It really should be a SN transformation. It's true that Earth matter can't change the decoherence but vacuum transformations could e.g. one of the mass states decays.
> 
> My impression was that `NoTransformation`, `CompleteExchange` and `ThreeFlavorDecoherence` are typically used as “black box transformations”—where researchers are interested in some other aspect and use these extreme scenarios to get a rough idea of how flavor transformations could affect studies of that other aspect. So it makes sense to me to keep these three separate?
> 
> However, if you think that any users might want to build a `TransformationChain` out of `ThreeFlavorDecoherence` in the SN plus additional transformations in vacuum/earth, then we could turn TFD into a SNTransformation. In that case, would `TransformationChain(in_sn=ThreeFlavorDecoherence())` automatically result in the same flavour equilibration that we get from the current implementation of TFD? If not, that might be very confusing.

The use case you describe for TFD is a useful one so my suggestion is to have a Phenomenological Transformation in __init__.py called e.g. **Equilibrate** which implements this case, and to have ThreeFlavorDecoherence as prescription in in_sn.py.  The ThreeFlavorDecoherence is useful for describing the effect of strong turbulence in supernovae; the TwoFlavorDecoherence describes the effect of weak turbulence.

jpkneller commented on python/snewpy/flavor_transformation/TransformationChain.py:
```diff
@@ -0,0 +1,71 @@
+        for t in self.transforms:
+            t.mixing_params = mixing_params
+        
+    def __call__(self, mixing_params):
+        """Convenience method: update mixing parameters and return self"""
+        self.set_mixing_params(mixing_params)
+        return self
+        
```
I agree with this suggestion. My presumption when reading a code is that member functions modify the existing class.

Sheshuk replied:
I also don't like this  `__call__` method - it is here only to mimic the previous behaviour, as in `__init__.py` we define the aliases:
https://github.com/SNEWS2/snewpy/blob/6ffae09ae4a95bae1115e7c0a351d60b629f2190/python/snewpy/flavor_transformation/__init__.py#L44-L56

so that previous code like `AdiabaticMSW(MixingParams)` would work.
I'm not sure about this whole interface - so I just tried to make a minimal adapter layer, to conform to the old interface. There are many other ways to do them, but I don't see any of them as simple and elegant. Make specific classmethod constructors, and use them as aliases in `init`?
Maybe you have an idea?

Sheshuk replied:
> but this wouldn’t work at all. (Instead, it would result in `msw_nufit == msw_pdg == AdiabaticMSW`.)
Could you explain, why? I thought, that in each line we create an instance of `AdiabaticMSW`, and then set the `_mixing_params` attribute for each of them specifically, so that `msw_pdg` and `msw_pdg` should be `AdiabaticMSW` instances with corresponding mixing parameters? Or am I missing something?

JostMigenda replied:
> so that previous code like AdiabaticMSW(MixingParams) would work.

That wasn’t quite the previous interface, though. Previously, [the function signature was](https://github.com/SNEWS2/snewpy/blob/f3150b454af94c52bc4720e4d1582a16d065474e/python/snewpy/flavor_transformation.py#L438) `def __init__(self, mix_angles=None, mh=MassHierarchy.NORMAL):` and we’re breaking this deliberately to pass in a `MixingParameters` instead.

So … I’d be okay with breaking that “even more” (at least for now—maybe we can discuss at the hackathon next month how an intuitive interface for that should look like?) and deleting the `__call__` for now before it causes any further confusion.

jpkneller commented:
Which branch should I start from? We seem to have a couple running in parallel.

JostMigenda commented:
Probably from the branch of this PR here? (`Sheshuk/FlavorTransformations_with_FlavorMatrices`) This shouldn’t conflict with the changes in #363.

Does that mean you’re done reviewing this PR and happy with it? If so, then I think the only remaining issue before we can merge this PR is the `__call__` question.

JostMigenda commented:
I’ve moved Jim’s commits to a separate branch (and separate PR #385). Hopefully, keeping that discussion separate will help us get this PR here merged sooner

jpkneller commented:
Did we decide on the new name for MSWEffect? 
My preference is for NumericalMSW.

Sheshuk commented:
@jpkneller, about your commit: the implementation of get_transformed_spectra is already done in #363 
So this MR here avoids this on purpose

jpkneller commented:
> @jpkneller, about your commit: the implementation of get_transformed_spectra is already done in #363 So this MR here avoids this on purpose

Sorry, I just saw the error and didn't check. Is there anything left to do with this PR?

jpkneller reviewed (approved):
Let's merge this branch and #363 and then deal with the remaining small issues.
